### ISSUE TITLE
Transient objects won't be GCed if their constructor open a resolution scope, leading to unexpected memory leak.

### ISSUE THREAD

**Issue by ryoasai** [dadhi/DryIoc#320]
I found that the current behavior of resolution scope is questionable and causes unexpected memory leak easily.

As the document mentions, I understand that this is according to the current spec,  but yet this behavior is unexpected for me when applied to transient scoped objects.

[ResoutionScope](https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/ReuseAndScopes.md#reusescopetotserviceservicekey)
> The code also implies that Foo itself will be automatically scoped to its scope. It may be important if you want to access Foo recursively from its dependency.
> 
> But with such an automatic scoping we still have a problem: how to dispose the scope.
> 
> In order to dispose the scope, DryIoc should somehow track the reference to it. This is exactly how it works, an automatically created scope will be held by either current open scope or by singleton scope. When the current scope or container with singletons is disposed - the automatic scope is disposed to.

Suppose you have a transient object which opens resolution scope for its constructor.

```C#
[OpenResolutionScope]
[Export, TransitentReuse]
public EditorViewModel {

    public EditorViewModel(DependencyA depA, DependencyB depB) {}
 
}

[Export, ResolutionScopeReuse]
public class DependencyA {}

[Export, ResolutionScopeReuse]
public class DependencyB {
    public DependencyB(DependencyA depA)
}

```
The intention of course is that the same instance of DependencyA is created and shared for resolving the root class. This works as expected for resolution, however, it causes EditorViewModel never deleted until global scope is disposed at application shutdown.

I wonder why the automatically created resolution scope need to be kept in singleton scope (or parent opened scope if any). Can you consider providing an option to automatically dispose temporary resolution scope? Or you should clarify this pit fall on documentation. Or you might be able to provide container option and Exception to warn unexpected memory leak according to DryIoc's  common design policy.

I think the above use case of using resolution scope on transient object is not rare.

**Comment by dadhi** [comment]
> I wonder why the automatically created resolution scope need to be kept in singleton scope (or parent opened scope if any). Can you consider providing an option to automatically dispose temporary resolution scope?

The problem is **when** to dispose of the scope then. 

I have cleanup-up the docs and added the section regarding [explicit scope disposal](https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/ReuseAndScopes.md#own-the-resolution-scope-disposal)

> Suppose you have a transient object which opens resolution scope for its constructor.

Yeah, for the **transient** service it is not working as you probably expect and it may be a problem :/
You would probably expect the object tree to look like this `MyTransient -> ScopeForItsSharedDependencies -> Deps...`
but it actually looks like this ` ScopeForItsSharedDependencies -> MyTransient -> Deps...`.

I am afraid the reason for this is technical - there is nothing to attach the scope to in `MyTransient` because the container does not have control for the User's object insides. But when you turn this around, the container is in control to open and attach `MyTransient` to the new scope.

Anyway, knowing the principle you may workaround it yourself like this (not pretty though):

```cs
public EditorViewModel {
    private EditorViewModel(DependencyA depA, DependencyB depB) {}

    [Export, TransitentReuse]
    public static EditorViewModel CreateEditorViewModel(IResolveContext r
        FactoryDelegate<DependencyA> depA, 
        FactoryDelegate<DependencyB> depB) 
    {
         using (var scope = r.OpenScope(EditorViewModelScope.Name)
             return new EditorViewModel(depA(scope), depB(scope))
    } 
}

[Export, CurrentScopeReuse(EditorViewModelScope.Name)]
public class DependencyA {}

[Export, CurrentScopeReuse(EditorViewModelScope.Name)]
public class DependencyB {
    public DependencyB(DependencyA depA)
}
```

**Comment by ryoasai** [comment]
Thank you for the workaround.

Actually, I could solve my problem by just changing the design of Dependency class and instead of passing resolution scoped object to constructor, I just pass it for a method from the root model object.

```C#
[Export, TransitentReuse]
public EditorViewModel {

    private DependencyA _depA;
    private DependencyB _depB;

    public EditorViewModel(DependencyA depA, DependencyB depB) 
    {
        _depA = depA;
        _depB =  debB;
    }
 
    public void SomeMethod()
    {
        _depB.process(_depA);
    }
}

[Export, TransitentReuse]
public class DependencyA {}

[Export, TransitentReuse]
public class DependencyB {
    public DependencyB()

    public processDependencyA(DependencyA depA) {
    }
}

```

I'm developing a WPF desktop application with Prism and I originally migrated from Unity container which has resolution scope. That's why I hit this problem. I think resolution scope is usually useful only for transient objects. By using resolution scope, I expected to share instance only during a single resolution process and any dependencies will be GCed with owner transient objects.

Singleton objects or regular scoped objects can just use longer life objects, so I suppose it's very rare that one time resolution scope is required for them.

It seems DryIoc originally had intrinsic resolution scope support but now it's implemented as regular scope mechanism. I guess the current behavior is something to do with this migration. 

If it's technically difficult to automatically dispose temporary resolution scope after resolution call, I'm afraid it's not very useful and even danger. You might well consider deprecating resolution scope support completely.

Anyway, I understand the current behavior follows the documentation, so it's not a bug. I don't request you to change the behavior. I posted this issue as a reminder and improvement suggestion.

**Comment by dadhi** [comment]
> If it's technically difficult to automatically dispose temporary resolution scope after resolution call, I'm afraid it's not very useful and even danger. You might well consider deprecating resolution scope support completely.

Maybe. There was not enouhgh feedback when I was shaping the feature, so currently it is hard to redesign it completely. But I found your case quite compelling and want to at least provide an explicit option for it.

I think the straightforward way is to add another option `avoidResolutionScopeTracking` and expose the respective attribute or attribute parameter.

**Comment by dadhi** [comment]
fixed via #323
